cros_gralloc/cros_gralloc_driver.cc

```diff
@@ -248,15 +250,29 @@ cros_gralloc_driver::cros_gralloc_driver()
 					video_idx);
 		}
 
+		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx)) {
+			drv_kms_ = drv_create(node_fd[virtio_node_idx]);
+			if (!drv_kms_) {
+				drv_loge("Failed to create driver for the virtio-gpu device with card id %d\n",
+                                          virtio_node_idx);
+				close(node_fd[virtio_node_idx]);
+				drv_kms_ = drv_render_;
+                        }
+		} else {
+			drv_kms_ = drv_render_;
+		}
+
 		// Init drv
 		DRV_INIT(drv_render_, gpu_grp_type, renderer_idx)
 		if (video_idx != renderer_idx)
 			DRV_INIT(drv_video_, gpu_grp_type, video_idx)
+		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx))
+			DRV_INIT(drv_kms_, gpu_grp_type, virtio_node_idx)
 	}
 
 	for (int i = 0; i < availabe_node; i++) {
 		free(node_name[i]);
-		if ((i != renderer_idx) && (i != video_idx)) {
+		if ((i != renderer_idx) && (i != video_idx) && ((drv_kms_ == drv_render_) && (i == virtio_node_idx))) {
```

Comment:
This line kind of confusing. Shouldn't it be clear like this: 
if ((i != renderer_idx) && (i != video_idx) && (i != virtio_node_idx)) {
and also in this circumstance, the virtio fd already closed in line 258

Reply:
in the case virtio_node_idx != renderer_idx
when we need use kms to allocate scanout buffer, shouldn't close the virtio_node_index. 
when we do not use kms to allocate scanout buffer, should close the virtio_node_index.
The line 258 you mentioned is the kms driver init fail case.
The condition ensure when the kms is not used to close virtio node.